nestkernel/stopwatch.h

```diff
@@ -0,0 +1,505 @@
+  }
+
+  bool
+  isRunning() const
+  {
+    bool isRunning = false;
+#pragma omp master
+    {
+      isRunning = walltime_timer_.isRunning();
+    };
+    return isRunning;
+  }
+
+  double
+  elapsed( timers::timeunit_t timeunit = timers::timeunit_t::SECONDS ) const
+  {
+    double elapsed = 0.;
+#pragma omp master
+    {
+      elapsed = walltime_timer_.elapsed( timeunit );
+    };
+    return elapsed;
+  }
+
+  void
+  reset()
+  {
+#pragma omp master
+    {
+      walltime_timer_.reset();
+      cputime_timer_.reset();
+    }
+  }
+
+  void
+  print( const char* msg = "",
+    timers::timeunit_t timeunit = timers::timeunit_t::SECONDS,
+    std::ostream& os = std::cout ) const
+  {
+#pragma omp master
+    walltime_timer_.print( msg, timeunit, os );
+  }
+
+  void
+  output_timer( DictionaryDatum& d, const Name& walltime_name, const Name& cputime_name )
+  {
+    def< double >( d, walltime_name, walltime_timer_.elapsed() );
+    def< double >( d, cputime_name, cputime_timer_.elapsed() );
+  }
+
+private:
+  timers::StopwatchBase< CLOCK_MONOTONIC > walltime_timer_;
+  timers::StopwatchBase< CLOCK_THREAD_CPUTIME_ID > cputime_timer_;
+};
+
+/** If the user deactivated detailed timers, Stopwatch instance with the detailed flag will become an empty Stopwatch,
+ * which will be safely ignored by the compiler, as if the instance was not declared (e.g. as a member).
+ */
+template <>
+class Stopwatch< StopwatchVerbosity::Detailed, StopwatchType::MasterOnly, std::enable_if< not use_detailed_timers > >
+{
+public:
+  void
+  start()
+  {
+  }
+  void
+  stop()
+  {
+  }
+  bool
+  isRunning() const
+  {
+    return false;
+  }
+  double
+  elapsed( timers::timeunit_t = timers::timeunit_t::SECONDS ) const
+  {
+    return 0;
+  }
+  void
+  reset()
+  {
+  }
+  void
+  print( const char* = "", timers::timeunit_t = timers::timeunit_t::SECONDS, std::ostream& = std::cout ) const
+  {
+  }
+  void
+  output_timer( DictionaryDatum&, const Name&, const Name& )
+  {
+  }
+};
+
+/** Only provide these template specializations if threaded timers are activated and always fall back to the base
+ * template specialization if not. Deactivate threaded detailed timers if all detailed timers are deactivated.
+ */
+template < StopwatchVerbosity detailed_timer >
+class Stopwatch< detailed_timer,
+  StopwatchType::Threaded,
+  std::enable_if_t< use_threaded_timers
+    and ( detailed_timer == StopwatchVerbosity::Detailed and not use_detailed_timers ) > >
+{
+public:
+  void
+  start()
+  {
+  }
+  void
+  stop()
+  {
+  }
+  bool
+  isRunning() const
+  {
+    return false;
+  }
+  double
+  elapsed( timers::timeunit_t = timers::timeunit_t::SECONDS ) const
+  {
+    return 0;
+  }
+  void
+  reset()
+  {
+  }
+  void
+  print( const char* = "", timers::timeunit_t = timers::timeunit_t::SECONDS, std::ostream& = std::cout ) const
+  {
+  }
+  void
+  output_timer( DictionaryDatum&, const Name&, const Name& )
+  {
+  }
+};
+template < StopwatchVerbosity detailed_timer >
```

Comment:
Insert an empty line above.

I get confused by how the selection of the different templates here works. Could you add an overview in the comment for the top-level `Stopwatch` template declaration about the approach to the selection logic?